Routing with virtualTree document (#3244)
SAP/luigi: docs/microfrontend-routing.md

<!-- meta
{
  "node": {
    "label": "Micro frontend routing",
    "category": {
      "label": "Advanced",
      "collapsible": true
    },
    "metaData": {
      "categoryPosition": 7,
      "position": 2
    }
  }
}
meta -->

# Micro frontend routing

> Luigi routing can be divided in two main parts: higher-level routing inside the main app (Luigi Core), and routing within the micro frontend (Luigi Client).

Reviewer: ```suggestion
Luigi routing can be divided in two main parts: higher-level routing of the main app (Luigi Core), and routing within the micro frontend (Luigi Client).
```

This document shows how you can handle the second way of routing while preserving the existing router of your micro frontend. The exact way this is achieved depends on the type of routing you use, but you should follow some general steps:

> 1. Enable [virtualTree](navigation-parameters-reference.md#virtualtree) in your Luigi configuration file.

Reviewer: ```suggestion
1. Enable [virtualTree](navigation-parameters-reference.md#virtualtree) in your Luigi configuration file for the node of your Microfrontend.
```

> 2. Add [Luigi Client](luigi-client-setup.md) to your micro frontend.

Reviewer: ```suggestion
2. Import [Luigi Client](luigi-client-setup.md) to your micro frontend.
```

> 3. Use the methods provided by Luigi Client's [linkManager](luigi-client-api.md#linkmanager) to update the Luigi Core URL.

Reviewer: ```suggestion
3. Use the methods provided by Luigi Client's [linkManager](luigi-client-api.md#linkmanager) to update the main app's URL and maintain navigation history.
```

        }, 
        
        routing: { 
            useHashRouting: true 
        }, 
        settings: { 
            responsiveNavigation: 'semiCollapsible',
            header: { 
                logo: 'img/luigi.png', 
                title: 'Svelte kit poc'
            }
        }
    });    
```

5. If you don't want to specify each subsequent navigation node in your application, you can use Luigi's [virtualTree](navigation-parameters-reference.md#virtualtree) feature.

## React routing

### Overview

This example provides general instructions on how to implement routing with React. The specifics might depend on which version of React you are using. 

### Steps 

1. In your Luigi configuration file, set [virtualTree](navigation-parameters-reference.md#virtualtree) to `true` for the node where you want to use React routing. This allows navigation to any of the node children's paths without the need for specifying them on the Luigi Core level.

```js
Luigi.setConfig({
  navigation: {
    nodes: [
      {
        hideSideNav: true,
        label: 'React',
        pathSegment: 'react',
        viewUrl: '//localhost:3000/',
        virtualTree: true,
      },
```

> 2. Inside your React app's `App.js` file, use Luigi's [linkManager](luigi-client-api.md#linkmanager) methods to make Luigi aware of the internal URL changes. Use [fromVirtualTreeRoot](luigi-client-api.md#fromvirtualtreeroot) to set routing to the parent route in the `virtualTree`, and [withoutSync](luigi-client-api.md#withoutsync) to disable Luigi Core handling of the URL. Finally, navigate to the correct path using Luigi Client's [navigate](luigi-client-api.md#navigate) method.

Reviewer: ```suggestion
2. Inside your React app's `App.js` file, use Luigi's [linkManager](luigi-client-api.md#linkmanager) methods to make Luigi aware of the internal URL changes.  This can be done in React by calling a LuigiClient navigation function inside your route change listener. As the route of your microfrontend changes, the LuigiClient navigation call will append it to the main apps route so that it is kept in the browser's session navigation history, allowing you to move back and forth while navigating internally. Use [fromVirtualTreeRoot](luigi-client-api.md#fromvirtualtreeroot) to set routing to the parent route in the `virtualTree`, and [withoutSync](luigi-client-api.md#withoutsync) to disable Luigi Core handling of the URL. Finally, navigate to the correct path using Luigi Client's [navigate](luigi-client-api.md#navigate) method. 
```

    "category": {
      "label": "Advanced",
      "collapsible": true
    },
    "metaData": {
      "categoryPosition": 7,
      "position": 2
    }
  }
}
meta -->

# Micro frontend routing

Luigi routing can be divided in two main parts: higher-level routing of the main app (Luigi Core), and routing within the micro frontend (Luigi Client).

This document shows how you can handle the second way of routing while preserving the existing router of your micro frontend. The exact way this is achieved depends on the type of routing you use, but you should follow some general steps:

1. Enable [virtualTree](navigation-parameters-reference.md#virtualtree) in your Luigi configuration file for the node of your micro frontend.
2. Import [Luigi Client](luigi-client-setup.md) to your micro frontend.
3. Use the methods provided by Luigi Client's [linkManager](luigi-client-api.md#linkmanager) to update the main app's URL and maintain navigation history.

Below, you can find some simple examples on how to keep routing in sync for different frontend technologies such as [Angular](#angular-routing), [SvelteKit](#sveltekit-routing), and [React](#react-routing).

<!-- accordion:start -->

> ## Angular routing

Reviewer 2: Here we should add a link to client-support-angular package with the info, that this should be used for the angular use case. Still we should keep the manual setup here for explanation, how it works under the hood.
